Android : new GameBrowser

BlisterB opened the pull request:
Hello everyone!

Here is the result of those two days of works : 
- I made a new GameBrowser for API > 15, hope you'll like it!
- Now we target API 23 but I managed to keep the min SDK to 10. I needed to adapt a lot of things but this sounds to works good ^^.
- I cleaned some part of code (this part is not over).
- I fixed some bug not reported.
- I've put a beautiful flashy setting button on the main activity, at the sight of everyone.

This pull request is not over, but I prefer to do this pull request to collect some feedbacks before finish it (I can't code for a week).

What remains to : 
- [ ] Deal with the new permission system of Android 6
<!> : If you are running on Android 6+, you have to manually give permission to the app (in Android Preferences). If you don't do this, the app cannot access to internal and external storage (don't forget we're dealing with Android platform hahaha).
- [ ] Do the "Share" feature
- [ ] Collect feedback and bugreport.
- [ ] Continue to clean the code.
- [ ] Refund the Settings Activity (with category and a uniform UI with the main activity)
- [ ] The .nomedia seems to not be added to new game directory

BlisterB commented:
Oh, we will probably have to slightly change the build script, to fit with the new target sdk.

Ghabry commented:
Can you make some screenshots of the new browser?

BlisterB commented:
Of course, here are some screenshots (if my mobile phone and github are compatible) : 
`
![tmp_24983-screenshot_20160711-003250248084068](https://cloud.githubusercontent.com/assets/2387475/16722052/519d7348-4744-11e6-82f7-26901db8607a.png)
`

BlisterB commented:
Of course, here are some screenshots : 
![]
(http://img11.hostingpics.net/pics/8211921363325510209917011263917879399914o.jpg)
![]
(http://img11.hostingpics.net/pics/98274113646988102099170113039181653436210o.jpg)
![]
(http://img11.hostingpics.net/pics/95231613639866102099170113439191556435259o.jpg)

Ghabry commented:
Could I have the new Android build line? :)

Ghabry commented on builds/android/app/src/main/java/org/easyrpg/player/DirectoryChooser.java:
```diff
@@ -201,7 +204,11 @@ public static String dirListToPath(LinkedList<String> l) {
+        if (!canRead) {
+           Log.e("isReadable", "Current context not allowed to read : " + path);
+        }
+        return canRead;
```
Please fix the whitespace everywhere (TAB for idendation)

Ghabry commented:
I installed on the server (in the toolchain dir, will push this to buildscripts when ready :D):
```
android update sdk -u -a -t android-23
android update sdk -u -a -t extra-android-support
```

Which is API23 and Android Support Lib

```
A problem occurred configuring project ':app'.
> Could not resolve all dependencies for configuration ':app:_debugCompile'.
   > Could not find com.android.support:appcompat-v7:23.2.1.
     Searched in the following locations:
         https://jcenter.bintray.com/com/android/support/appcompat-v7/23.2.1/appcompat-v7-23.2.1.pom
         https://jcenter.bintray.com/com/android/support/appcompat-v7/23.2.1/appcompat-v7-23.2.1.jar
     Required by:
         android:app:unspecified
   > Could not find com.android.support:design:23.2.1.
     Searched in the following locations:
         https://jcenter.bintray.com/com/android/support/design/23.2.1/design-23.2.1.pom
         https://jcenter.bintray.com/com/android/support/design/23.2.1/design-23.2.1.jar
     Required by:
         android:app:unspecified
   > Could not find com.android.support:cardview-v7:23.2.1.
     Searched in the following locations:
         https://jcenter.bintray.com/com/android/support/cardview-v7/23.2.1/cardview-v7-23.2.1.pom
         https://jcenter.bintray.com/com/android/support/cardview-v7/23.2.1/cardview-v7-23.2.1.jar
     Required by:
         android:app:unspecified
   > Could not find com.android.support:recyclerview-v7:23.2.1.
     Searched in the following locations:
         https://jcenter.bintray.com/com/android/support/recyclerview-v7/23.2.1/recyclerview-v7-23.2.1.pom
         https://jcenter.bintray.com/com/android/support/recyclerview-v7/23.2.1/recyclerview-v7-23.2.1.jar
     Required by:
         android:app:unspecified
```

ItsSchoko commented:
Hey there, good to see that someone is going the same way. :D 

I've just made a short demo clip of the current status:
[https://youtu.be/-U3OpDch92E](https://youtu.be/-U3OpDch92E)

For some inspiration, I also add my mockup graphic which I created before programming the UI. (Graphic is not up to date anymore)

![easy_rpg_player_ui_overview_2](https://cloud.githubusercontent.com/assets/17786862/16734463/e8c7f356-4785-11e6-96d1-0521b82a6c07.jpg)


The UI itself is minSDK10 and maxSDK23, had some minor problems like white backgrounds and stuff, because 4.4 doens't inherit the background.

I also plan to make a second list view option for larger screens (e.g. TV), like in your screenshots.

And as always, a DropBox link to the Screenshots of the Mockup, inside the APK folder is an APK of the current status.
[https://www.dropbox.com/sh/0zqznc8x7v7sar5/AACeNLToHxLgHn2vZX_fPzIka/UI](https://www.dropbox.com/sh/0zqznc8x7v7sar5/AACeNLToHxLgHn2vZX_fPzIka/UI)

BlisterB commented:
@Ghabry : I've updated the dependencies following this stack : 
http://stackoverflow.com/questions/29284236/failed-to-resolve-com-android-supportappcompat-v722-and-com-android-supportre

@ItsSchoko : Your UI is nice, I'm curious to see the final result, especially for the settings activity (I don't have the time to redo it).
Nevertheless I don't really like the new Button Layout, I like the current one which stay discreet.

BlisterB commented:
I had the time to watch the video. This is really nice.
Considering this work, I'll cut some functionality for this Pull Request and just do the permission asking. This GameBrowser will be avaible until @ItsSchoko will finish his own, to decide which one will be our futur base work.

ItsSchoko commented:
@BlisterB I'm curious too, it's still a long way to go.
This Button Layout is part of some extra button sets I've created, I'm just to lazy to use the old ones inside the mockup project. (There are also flat discreet ones)
My Plan is to rework the Button Layout Editor, which then should work with .svg files, color change, scaling, etc.

A lot of comments on the Playstore just scream for better usability, which I try to reach.
I know a lot of you guys are busy with core features, so I decided to help in creating a better enduser UI experience with my first app.

Ghabry commented:
@BlisterB 
https://gradleplease.appspot.com/
Could you try this settings? (change the file, do a "git commit -a --amend" and "git push -f") ;)

I installed now also the 2 maven repos, no change.

BlisterB commented:
Yeah until recently we were unable to use external dependencies so it was impossible to use something API > 10 and it was difficult to do beautifull things. And as I am not student anymore my time si shorter :/.

Your UI is really nice, but please pay attention with the dependencies that we can found on the web : they must be compliant with our licences (which is a beautifull bottleneck hahaha).

@Ghabry : Okay!

ItsSchoko commented:
@BlisterB Already got that coverd.

- the Navigation Menu icons are Google system icons [https://design.google.com/icons/index.html](https://design.google.com/icons/index.html)
- the game icons are just there to illustrate what you will see, so only content from the games you have "installed" are loaded into the list.
- and rest is my own creation (mockup phones, buttons sets etc.) or from EasyRPG

BlisterB commented:
Okay, sounds cool!

@Ghabry : I did what you've asked :).

Ghabry commented:
@BlisterB 
Do I have to install Maven and download the support library dependencies manually? Because the repository folder is almost empty.

BlisterB commented:
Nope, everything works with gradle, 
I think you're right, this is what seems to tell this PR : http://stackoverflow.com/questions/36222111/android-could-not-find-com-android-supportsupport-v423-2-1 .

Ghabry commented:
Jenkins: Test this please

Ghabry commented:
@BlisterB 
Fixed on the server now :)

Ghabry commented:
One obvious problem of your Title folder scanning is: It will pick the wrong one when there is more then 1 file in it... But to get the correct one you would need to parse the RPG_RT.ldb :D

BlisterB commented:
Ghabry : yeeees the title screen is very dumb I agree but it would be a mess to integrate ldb parsing and terrible for performances :p.
I'll do the select layout feature for the game menu :).

BlisterB commented:
@Ghabry Now the app ask for storage permission (this thing is awfull to handle in android code).
For the edit button in the select layout, I checked and this is always not possible to do this quickly without creating a whole new class.
Because I won't be free for a week I think I'm over for this PR and I add the things I didn't in the Android to do list thread.
Of course if something bother you don't hesitate to tell me :).

Ghabry commented:
Okay I will keep this PR on hold then until I have time to do more testing. Pushing this directly to the Android users is a bit risky because it's workweek :D

BlisterB commented:
Hi everyone, I'm back with an Java 8 OCA in my pocket hihihi ^^.

@Ghabry Do you have suggestions for this PR?

Ghabry commented:
@BlisterB 
Imo the game title graphics in your tile view should depend on the screen size... Currently gigantic and small displays get 2 games per row. And the aspect ratio isn't correct (title graphics are cut, not the whole 320x240 graphic is rendered).

But my bigger worry is that some users will probably hate this view and show all there hate with 1 star ratings!
Maybe you could add a button to toggle between tile and list view? Basicly like in the mockup of Schoko. Small title graphic on the left, followed by game title. Is more space friendly.
But really depends on who you ask, both view options make sense.

BlisterB commented:
I agree with you with for the switch layout button.
For the moment it's pretty cool with few games in portrait mode, but that would me to heavy with more than 5 games. I also like the Schoko mockup.
@ItsSchoko where is the advancement of your work, do you have any idea of when it will be ready for a pull request? I don't really want to do a double work since I found your result better.

Ghabry commented:
@BlisterB 
I think your GUI changes are optimal for getting initial feedback from the PlayStore users. I see yours as a perfect gap filler until Schokos GUI is ready. And you already wrote some useful code for getting the title graphic e.g.

Lets scratch the list layout part because Schoko already did it but is it possible without too much work to make the tile view resolution aware? To render more stuff per row on larger displays? Because Schokos Ui doesn't offer this view it would be a nice addition.
After this being added I think we can approve this PR and push it to the store :+1: and wait for angry Samsung users :laughing:

BlisterB commented:
That was in my initial plans to improve the layout in further PR, but this isn't difficult, I'll just have to do some tests with different screen size VM :).

BlisterB commented:
I adapted the grid layout depending on the horizontal size. A thumbnail now has a minimum size of 200dpi, I just have a smartphone so I can't test on a tablet. Can someone tell me I I should slightly minify the thumbnail size ?

Ghabry commented:
Ignore the build failure, is our fault.
Will test later.

Ghabry commented:
Jenkins: Test this please

Ghabry commented:
You can test using an emulator. ;)

I just tested it on my smartphone and it worked, will have access to a tablet in the evening.

Could you give the top-pane with the EasyRPG text a green color? Blue doesn't fit at all in our corporate design \*g\*

BlisterB commented:
I changed the color to match with the toolbar :).

Ghabry commented:
Uhm now everything is blue instead of green.

Anyway some screens from my tablet:
![screenshot_2016-07-29-23-17-33](https://cloud.githubusercontent.com/assets/1331889/17263697/bc3e1566-55e2-11e6-9d02-77a194eb09e4.png)

![screenshot_2016-07-29-23-17-10](https://cloud.githubusercontent.com/assets/1331889/17263707/ca4204ce-55e2-11e6-842e-303a1d483bae.png)

An obvious problem is the settings button. The text flows under it. That looks bad. Any idea how to solve this?

Ghabry commented:
The heuristic also works surprisingly well if the game does not contain the rtp title graphics. And yume nikki uses xyz files

BlisterB commented:
Happy for the xyz format ^^.
Apologies for the blue thing, I did a mistake reading your post huhuhu. I'll fix it tomorrow, and move the preferences button in the toolbar :).

Ghabry commented:
You read wrong again. The global button (pink circle) is fine. I mean the per-game settings.
See Vampires Dawn

BlisterB commented:
I definitively shouldn't try to work at midnight hahaha.
I fixed the text overlapping problem and I changed the color scheme, I also set the pink color to the complementary of the blue in the toolbar : yellow.

carstene1ns commented:
![current_colors](https://p.f4ke.de/gaBiZG0S7d.png)
I say it again, our color is GREEN, not BLUE.

BlisterB commented:
Wow, I must had a unconscious blockage against the green color, or a color blindness condition.
This is now all GREEN.

Ghabry commented:
Good. Now I ask you to remove the redundant color commits.
Open a shell and go into the easyrpg directory and run

```
git rebase -i HEAD~5
```

You will get an editor listing with the last 5 commits, all set to "pick" in the first column.

Edit it so that the 2nd and the 3rd commit are changed to "fixup" like in this screenshot:
![screenshot_20160730_131739](https://cloud.githubusercontent.com/assets/1331889/17270007/32d4823a-5658-11e6-90fb-311e444994c9.png)

Save (in nano strg+O) and exit (Strg+X)

Now run ```git log``` and confirm that the commits are gone.

Finally do a forced push ```git push -f```

Ghabry commented:
@BlisterB  stop it, it's correct now

Ghabry commented:
![xxx](https://cloud.githubusercontent.com/assets/1331889/17270721/4fbf4e7c-566a-11e6-9335-ee4809b32ae0.png)

BlisterB commented:
Hahaha

Ghabry commented:
![screenshot_2016-07-30-16-05-18](https://cloud.githubusercontent.com/assets/1331889/17270997/e6c99e76-566f-11e6-976d-256bf5dfbbe5.png)

Ghabry commented on builds/android/app/src/main/res/values-w820dp/dimens.xml:
```diff
@@ -0,0 +1,6 @@
+    <!-- Example customization of dimensions originally defined in res/values/dimens.xml
+         (such as screen margins) for screens with more than 820dp of available width. This
+         would include 7" and 10" devices in landscape (~960dp and ~1280dp respectively). -->
+    <dimen name="activity_horizontal_margin">64dp</dimen>
```
You overwrite here the horizontal_margin from 16dp to 64dp for w820dp this could be a hot candidate for the issue.

BlisterB replied:
Hmm this is an auto-generated file by Android Studio, if you change it, does it solve the problem?
Because this problem doesn't occur when you rotate your device, I tend to think that the line 164 in GameBrowserActivityAPI15.java is the source of the problem:
`int nbGamePerRow = configuration.screenWidthDp / THUMBNAIL_HORIZONTAL_SIZE_DPI;`

carstene1ns commented on builds/android/app/src/main/res/values/strings.xml:
```diff
@@ -104,4 +104,8 @@ Please tell us in detail what went wrong.\n
+
+    <!--GameBrowserAPI15-->
+    <string name="navigation_drawer_open">Open navigation drawer</string>
+    <string name="navigation_drawer_close">Close navigation drawer</string>
```
Please fix the white spaces here to make it <TAB> like the rest.

BlisterB replied:
Oh, I didn't saw the problem because it's indented with 4 space, and the other with tab (no difference in my IDE).
Because 4 space is the Java standard (which is stupid imho), I will try to keep this way to indent.

BlisterB commented:
@Ghabry does my last commit solved your problem?
It seems this way to obtain the screen width is better than using the Configuration object.

Ghabry commented:
Not fixed. Maybe you could test it in an emulator?

BlisterB commented:
Y-e-s I'm using multiple VM to test the GameBrowser. But I didn't faced this problem.
Could you give me some information about your model (screen size, android version, eventually the manufacturer if a friend have a similar model), or at least tell me if your fix (the file deletion) worked?

Please remember that Android platform is a mess, possibly every Android version/model could give a slightly different result.

Ghabry commented:
Oh, okay :D
It's a Android 4.1 Samsung(!) Galaxy Note.

Size 	8.0 inches (~64.8% screen-to-body ratio)
Resolution 	800 x 1280 pixels (~189 ppi pixel density)

BlisterB commented:
I succeed in reproducing the layout bug, please excuse me for the long waiting time, your fix was the right solution @Ghabry.
I also deleted the previous commit which became useless.

Ghabry commented:
Jenkins: Test this please

Ghabry commented:
In my opinion this can be merged now. Anybody with further criticism?

carstene1ns commented:
I just hope transifex does not burp on the whitespace change, otherwise ok.

Ghabry commented:
Burped a bit. @BlisterB , any idea where these localizations are coming from? :o

![screenshot_20160807_001332](https://cloud.githubusercontent.com/assets/1331889/17459356/2435ce50-5c34-11e6-9eda-9f6606b3f8f7.png)

BlisterB commented:
Wtf, where does this come from oO.

Ghabry commented:
![wtfandroid](https://cloud.githubusercontent.com/assets/1331889/17459364/8ca8cb22-5c34-11e6-90af-49acaa22afde.jpg)

Ghabry commented:
As @carstene1ns guessed the languages are part of the appcompat lib. I just looked in one of the appcompat files with nano and found references to values-[langcode] files that match the ones on the screenshot.

BlisterB commented:
Hmm, is this something severe?

Ghabry commented:
No, we agreed on the chat to ignore it. Too much work to "solve".

BlisterB commented:
Okay

BlisterB commented:
I'm currently cleaning the Settings Activity, I'm pretty ashamed to have done something so dirty huhuhu.

Ghabry commented:
APK is published. I will forward all crash reports here tomorrow. Good night.

Ghabry commented:
@BlisterB 
Users are happy. no feedback at all, neither positive nor negative. So must be good when nobody complains ;)
Also no crashes received.

BlisterB commented:
This is a good news ^^.
I intend to improve the code related to Preferences Activity which is actually awfull, and do an improvment of the portrait mode inspired by the ItsSchoko mock up (because the actual layout take obviously too much space). Before starting the study for the Oracle Java OCP huhuhu.

Ghabry commented:
You can also think about how to make it more obvious how to edit layouts because we found a user who couldn't figure out to press the "back button" to open the menu ;)

BlisterB commented:
lol

Ghabry commented:
@BlisterB 
A small update about how the people like the Game Browser: By now still no feedback (and no crashes) but our average rating increased over the last weeks and somehow most ratings are now 4 and 5 stars. I can't say that this is a correlation with your browser (we also had many other game fixes) but at least the amount of stupid questions reduced, looks like most people find the settings now.

One person even changed her rating to 5 stars because they were happy that adding shift is possible now m(
